### ISSUE TITLE
Add a run id that wires builds together

### ISSUE THREAD

**Issue by evanchaoli** [concourse/concourse#3905]
## What challenge are you facing?

A pipeline may contain multiple jobs. A typical uses is like, job1 is triggered by a git push (or something else), which generates build1, once build1 passes, job2 is triggered, which generates build2, and so on...

Logically, here build1 and build2 belong to a single CI process, we want to evaluate data based on an entire pipeline run, including all builds involved in the run.

I don't find anything that can wire build1 and build2 as a pipeline run, we can only retrieve individual build's info.

## What would make this better?

The following is my proposal for discussion and comment:

1. Add a run_id, an auto generated integer, to build table.
2. Once a build is triggers, generate a run_id and assign to it.
3. If a build passes and triggers following builds, assign its run_id to the following builds.

So that by querying run_id, we can get all builds of a pipeline run, therefore evaluate metrics at pipeline level. 

## Are you interested in implementing this yourself?

If my proposal is accepted, my team is willing to contribute this feature.

**Comment by pivotal-jamie-klassen** [comment]
If your jobs are are this tightly coupled, it kinda sounds like they might more logically be the same job? What would be your concerns about just concatenating their plans together so that they run as a unit?

**Comment by pivotal-jwinters** [comment]
@pivotal-jamie-klassen in the past I've split a job into multiple jobs for better visibility in the UI. So I don't have to click on the build to know which step failed.

@evanchaoli this is actually something we've been thinking about https://github.com/concourse/concourse/issues/3602#issuecomment-479541770. 

Although we've been approaching it from a model all connections between jobs (instead of modelling a  pipeline run itself). We were going to run a query to build the pipeline run from the connections. It would be interesting to dig into the differences between these two approaches.

/cc @clarafu @vito

**Comment by vito** [comment]
Yeah this sounds a lot like causality (https://github.com/concourse/concourse/issues/1538).

**Comment by evanchaoli** [comment]
I have gone through #1538 and #3602, however not seeing a result. @vito suggested a query in #1538, however I don't see how a concourse user can make the query? Do we expect concourse user to query the database directly?

Most of concourse users just came in from Jenkins as well as other existing CI tools. I have seen many of users on our concourse cluster have their pipelines like this:

    Trigger (git) -> Build Job -> Test Job -> Deploy Job

A lot of users have been asking me that how can they evaluate their CI runs from pipeline perspective. They want to evaluate execution time, success rate etc from pipeline perspective.

Here I can see two solutions:

* Re-educate users to combine multiple jobs into a single job. A job in fact reflects to a pipeline in other CI tools.
* As I proposed, add a run_id to every build, use run_id to wire builds together. If run_id is implemented, Concourse may provide a view of Runs. Run_id will avoid such confusions caused by resource version, such as how to handle re-trigger. Each run should have a unique run_id.

**Comment by vito** [comment]
@evanchaoli We don't expect users to be running queries against the database. This is a feature that we've spiked on and not yet implemented. There are screenshots in #1538 showing the progress we made on the spike.

**Comment by StevenArmstrong** [comment]
Just a user perspective here. Concourse if you place ci, test and deploy all under the 1 job you will lose the lovely visualisation and the individual re-run ability that you currently have at job level. Tasks are a poor substitute for jobs.
Job visualisation is one of the main reasons that many users choose concourse so I don't think asking users to place everything under 1 job with lots of tasks is a viable solution. 

I think a way of grouping jobs together into a pipeline flow with end to end execution tracking with a pipeline_id would be a great implementation. Instead of trying to re-invent the wheel I think concourse needs to really bite the bullet and model a way of having a traditional pipeline flow for users if they need it.

So as a simple example a user has job 1, job 2 and job 3

When job 1 is triggered by a resource change it generates the new pipeline flow for job 1 so job1, job 2 and job 3 move to a grey colour to show they are pending a new pipeline flow run. Job 1 will show grey and running. When job 1 completes it will go green or red and  generate a pipeline_id that can then passed to job 2 and 3 and will radiate green when complete or fail at job 2 or 3.

Users such as myself can then report off of the pipeline_id as it chains runs together.
Concourse could show the pipeline run as a historical view based on where the pipeline_id got to so show a run, failed, not run status.

This construct could be done using a pipeline_flow tag in yaml applied to a job or even a pipeline flow resource so any number of jobs can be chained together.

This feature could give users through the gui the ability to see where a current pipeline flow and an easy to understand radiator of the last run. As if jobs werent run in a pipeline flow they would show a new grey status showing not run.
You can also model multiple pipeline_flows within the overall pipeline as any amount of jobs can be tagged on a flow. A 1st implementation could make this 1-1 for ease of use or use a resource.

I see that this implementation wouldn't break any constructs and would give a lot of users what they are missing.

At the moment the main criticisms we see for concourse are I don't know at what stage my deployment is at and how all my jobs ran in the previous pipeline flow run. Also users want to track kpis on the end to end flow such as timings and this isn't easy at the moment as we have no pipeline_id object. These issues could all be solved by having clear defined pipeline flows. 
This is the number 1 feature our product manager gets from our teams.

I appreciate that you guys like to be requirement focussed but I feel looking at the rfcs you are kind of ignoring the lessons learned by Jenkins and Gocd that already have solved this issue. So I would ask if you could consider this feature request it would be very helpful to your users.

**Comment by vito** [comment]
@StevenArmstrong I've had similar thoughts as a way to eliminate use of the `pool` resource just to enforce sequential 'pipeline runs'. I'd also like it as a way to eliminate `serial_groups` as people often mis-use it expecting it to have 'pipeline run' semantics. Never had enough context at once to put together a proposal, though.

We haven't had the acuteness of experience with the issue required to write an RFC for it ourselves - RFCs take up a lot of mental bandwidth. Would you be willing to put one together? It sounds like a significant proposal that would need to be mapped out before we can just 'do' it.

**Comment by StevenArmstrong** [comment]
Yeah definetly we are using concourse heavily at just eat and are big fans
so I didn't want the above to sound too negative. I can work with the team
at just eat to pull one together with all the details and pass it on for
your review. Do you have an up to date template I can use? @vito

**Comment by vito** [comment]
@StevenArmstrong Sounds great! The [example RFC template](https://github.com/concourse/rfcs/tree/master/000-example) is up-to-date. The format is flexible though so feel free to add or remove sections as needed.

**Comment by pivotal-jamie-klassen** [comment]
Pretty sure this is still topical. It's not immediately obvious to me if this can be rolled into any of the [v10 dreams](https://blog.concourse-ci.org/core-roadmap-towards-v10/) that you've described recently, but @vito maybe you can see a connection that i cannot.

**Comment by vito** [comment]
@pivotal-jamie-klassen It's pretty orthogonal to v10 as that's all 'Core' focused - I think the remaining work for this is all in UX/API.

The latest development on this front is that the new algorithm resulted in the addition of a database table which will make it a lot easier to trace pipeline flows (`build_pipes`).

But there's currently no 'next step' - I've been focusing on Core all this time. We could work this into our UX roadmap but I'm not sure when we'd get to it. UX track is fairly small and split between RelEng, and honestly build page performance is probably more pertinent.

I'll just label this with 'help wanted' for now so the stale bot doesn't kill it. If anyone's interested in tackling this after v6.0 when the new algorithm is in place, let us know!

**Comment by evanchaoli** [comment]
@vito As I raised this issue, and our users are waiting for this feature. I'd like to help on this once the new algorithm is in place.

**Comment by vito** [comment]
@evanchaoli Great! For what it's worth, I finally slotted this into our [roadmap](https://github.com/concourse/concourse/projects/21) - but it's pretty far down on the UX track. So feel free to pick it up since it'll probably get done faster that way.

**Comment by vlfig-iov42** [comment]
Doesn't this go against the grain of causality? It sounds like the use case is to get information about a series of builds that belong to a certain "process". I wonder if this isn't subsumed by the idea, provided by causality, of getting a view of (the transitive closure of) all the builds that "went into" a certain build or resource revision.
